Detect borrow error involving sub-slices and suggest `split_at_mut`

estebank opened the pull request:
```
error[E0499]: cannot borrow `foo` as mutable more than once at a time
  --> $DIR/suggest-split-at-mut.rs:13:18
   |
LL |     let a = &mut foo[..2];
   |                  --- first mutable borrow occurs here
LL |     let b = &mut foo[2..];
   |                  ^^^ second mutable borrow occurs here
LL |     a[0] = 5;
   |     ---- first borrow later used here
   |
   = help: use `.split_at_mut(position)` or similar method to obtain two mutable non-overlapping sub-slices
```

Address most of #58792.

For follow up work, we should emit a structured suggestion for cases where we can identify the exact `let (a, b) = foo.split_at_mut(2);` call that is needed.

estebank committed:
Detect borrow error involving sub-slices and suggest `split_at_mut`

```
error[E0499]: cannot borrow `foo` as mutable more than once at a time
  --> $DIR/suggest-split-at-mut.rs:13:18
   |
LL |     let a = &mut foo[..2];
   |                  --- first mutable borrow occurs here
LL |     let b = &mut foo[2..];
   |                  ^^^ second mutable borrow occurs here
LL |     a[0] = 5;
   |     ---- first borrow later used here
   |
   = help: use `.split_at_mut(position)` or similar method to obtain two mutable non-overlapping sub-slices
```

Address most of #58792.

For follow up work, we should emit a structured suggestion for cases where we can identify the exact `let (a, b) = foo.split_at_mut(2);` call that is needed.

fee1-dead commented on tests/ui/borrowck/borrowck-overloaded-index-autoderef.stderr:
```diff
@@ -17,6 +17,8 @@ LL |     let q = &mut f[&s];
    |                  ^ second mutable borrow occurs here
 LL |     p.use_mut();
    |     - first borrow later used here
+   |
+   = help: use `.split_at_mut(position)` or similar method to obtain two mutable non-overlapping sub-slices
```
This seem to be a regression. We should check that the indexing is actually done on a slice and not a user custom type.

estebank replied:
I can easily check whether we have a `[]` or `Vec` and be more accurate that `split_at_mut` will work. We can also look for all methods on any custom type that being indexed on that returns `(&mut T, &mut T)` or `(&mut T, T)`. I was hoping not doing that as part of this PR, but I can.

If the comment instead means "they are doing `p.use_mut()`, where's the indexing, it happens earlier.

I am of the opinion that any type that supports indexing should have a method akin to `split_at_mut`. We should likely customize the error for types from the local crate, as *those* are much less likely to have an analogue.

estebank committed:
Only suggest `split_at_mut` on indexing borrowck errors for std types

estebank committed:
Mention `split_at_mut` when mixing mutability in indexing ops

Emit suggestion when encountering

```rust
let a = &mut foo[0];
let b = &foo[1];
a.use_mut();
```

estebank committed:
Don't suggest `split_at_mut` when the multiple borrows have the same index

estebank committed:
Check equivalence of indices in more cases

fee1-dead commented on compiler/rustc_hir/src/hir.rs:
```diff
@@ -1811,6 +1811,41 @@ impl Expr<'_> {
         }
     }
 
+    /// Whether this and the `other` expression are the same for purposes of an indexing operation.
+    ///
+    /// This is only used for diagnostics to see if we have things like `foo[i]` where `foo` is
+    /// borrowed multiple times with `i`.
+    pub fn equals(&self, other: &Expr<'_>) -> bool {
```
The naming seems confusing as it doesn't specify that this is only in terms of indexing operations. Perhaps rename to `equivalent_for_indexing`?

fee1-dead reviewed (changes requested):
One small nit, otherwise looks good to me, thanks! r=me.

bors commented:
:umbrella: The latest upstream changes (presumably #124136) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

estebank committed:
review comment: rename method

estebank commented:
@bors r=fee1-dead

bors commented:
:pushpin: Commit 64a4cdcfd4db72b7fc9b24521111f4f0edac3531 has been approved by `fee1-dead`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=fee1-dead 64a4cdcfd4db72b7fc9b24521111f4f0edac3531 -->
<!-- homu: {"type":"Approved","sha":"64a4cdcfd4db72b7fc9b24521111f4f0edac3531","approver":"fee1-dead","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
